### ISSUE TITLE
Problematic Markdown list formatting

### ISSUE THREAD

**Issue by kareldonk** [prettier/prettier#16929]
Prettier version > 3.4.0 introduced a change in Markdown list formatting that is different with previous behavior and causes alignment inconsistencies when multiple paragraphs are used in lists.

**Input:**

```
Here is a test list:

-   Lorem Ipsum is simply dummy text of the printing and typesetting industry.
    Lorem Ipsum has been the industry's standard dummy text ever since the 1500s,
    when an unknown printer took a galley of type and scrambled it to make a type
    specimen book. It has survived not only five centuries, but also the leap into
    electronic typesetting, remaining essentially unchanged. It was popularised in
    the 1960s with the release of Letraset sheets containing Lorem Ipsum passages,
    and more recently with desktop publishing software like Aldus PageMaker
    including versions of Lorem Ipsum.

    It is a long established fact that a reader will be distracted by the readable
    content of a page when looking at its layout. The point of using Lorem Ipsum
    is that it has a more-or-less normal distribution of letters, as opposed to
    using 'Content here, content here', making it look like readable English. Many
    desktop publishing packages and web page editors now use Lorem Ipsum as their
    default model text, and a search for 'lorem ipsum' will uncover many web sites
    still in their infancy. Various versions have evolved over the years,
    sometimes by accident, sometimes on purpose (injected humour and the like).

-   Lorem Ipsum is simply dummy text of the printing and typesetting industry.
    Lorem Ipsum has been the industry's standard dummy text ever since the 1500s,
    when an unknown printer took a galley of type and scrambled it to make a type
    specimen book. It has survived not only five centuries, but also the leap into
    electronic typesetting, remaining essentially unchanged. It was popularised in
    the 1960s with the release of Letraset sheets containing Lorem Ipsum passages,
    and more recently with desktop publishing software like Aldus PageMaker
    including versions of Lorem Ipsum.
```

**Output in version 3.4.0:**

```
Here is a test list:

- Lorem Ipsum is simply dummy text of the printing and typesetting industry.
  Lorem Ipsum has been the industry's standard dummy text ever since the 1500s,
  when an unknown printer took a galley of type and scrambled it to make a type
  specimen book. It has survived not only five centuries, but also the leap into
  electronic typesetting, remaining essentially unchanged. It was popularised in
  the 1960s with the release of Letraset sheets containing Lorem Ipsum passages,
  and more recently with desktop publishing software like Aldus PageMaker
  including versions of Lorem Ipsum.

    It is a long established fact that a reader will be distracted by the readable
    content of a page when looking at its layout. The point of using Lorem Ipsum
    is that it has a more-or-less normal distribution of letters, as opposed to
    using 'Content here, content here', making it look like readable English. Many
    desktop publishing packages and web page editors now use Lorem Ipsum as their
    default model text, and a search for 'lorem ipsum' will uncover many web sites
    still in their infancy. Various versions have evolved over the years,
    sometimes by accident, sometimes on purpose (injected humour and the like).

- Lorem Ipsum is simply dummy text of the printing and typesetting industry.
  Lorem Ipsum has been the industry's standard dummy text ever since the 1500s,
  when an unknown printer took a galley of type and scrambled it to make a type
  specimen book. It has survived not only five centuries, but also the leap into
  electronic typesetting, remaining essentially unchanged. It was popularised in
  the 1960s with the release of Letraset sheets containing Lorem Ipsum passages,
  and more recently with desktop publishing software like Aldus PageMaker
  including versions of Lorem Ipsum.
```

**Expected output:**

```
Here is a test list:

-   Lorem Ipsum is simply dummy text of the printing and typesetting industry.
    Lorem Ipsum has been the industry's standard dummy text ever since the 1500s,
    when an unknown printer took a galley of type and scrambled it to make a type
    specimen book. It has survived not only five centuries, but also the leap into
    electronic typesetting, remaining essentially unchanged. It was popularised in
    the 1960s with the release of Letraset sheets containing Lorem Ipsum passages,
    and more recently with desktop publishing software like Aldus PageMaker
    including versions of Lorem Ipsum.

    It is a long established fact that a reader will be distracted by the readable
    content of a page when looking at its layout. The point of using Lorem Ipsum
    is that it has a more-or-less normal distribution of letters, as opposed to
    using 'Content here, content here', making it look like readable English. Many
    desktop publishing packages and web page editors now use Lorem Ipsum as their
    default model text, and a search for 'lorem ipsum' will uncover many web sites
    still in their infancy. Various versions have evolved over the years,
    sometimes by accident, sometimes on purpose (injected humour and the like).

-   Lorem Ipsum is simply dummy text of the printing and typesetting industry.
    Lorem Ipsum has been the industry's standard dummy text ever since the 1500s,
    when an unknown printer took a galley of type and scrambled it to make a type
    specimen book. It has survived not only five centuries, but also the leap into
    electronic typesetting, remaining essentially unchanged. It was popularised in
    the 1960s with the release of Letraset sheets containing Lorem Ipsum passages,
    and more recently with desktop publishing software like Aldus PageMaker
    including versions of Lorem Ipsum.
```

**Why?**

Consistent alignment within lists with multiple lines and paragraphs, or lists with items containing code blocks or blockquotes etc.

**Comment by MartenBE** [comment]
In short, https://github.com/prettier/prettier/pull/15526 was introduced to prevenv `-.` expanding to `-...` as it didn't made sense in a lot lists. This works fine, but as discussed in the thread of the PR, the case with multiple paragraphs is a difficult edge case.
My proposal would be to implement something that the first `-.` or `-...` determines what you want with the rest of the list.

**Comment by br3ndonland** [comment]
### An admonition about admonitions

The update from https://github.com/prettier/prettier/pull/15526 introduces additional problematic behavior when formatting Markdown documents containing admonitions. It seems to be related to the original post _(list item -> empty line -> indented content)_.

Admonitions are supported by [Docusaurus](https://docusaurus.io/docs/markdown-features/admonitions), [Python-Markdown](https://python-markdown.github.io/extensions/admonition/), and other projects that use Python-Markdown like [Material for MkDocs](https://squidfunk.github.io/mkdocs-material/reference/admonitions/). The suggestion for formatting admonitions with Prettier is to add an empty line between the admonition directive and the admonition content ([Docusaurus docs](https://docusaurus.io/docs/markdown-features/admonitions#usage-with-prettier), https://github.com/prettier/prettier/issues/8640).

Given this `.prettierrc` ([Python-Markdown requires 4 space indentations](https://python-markdown.github.io/#differences)):


```json
{
    "tabWidth": 4
}
```

This Markdown file `example.md` formatted by Prettier 3.3:

````
# Example

-   This is an unordered list.
    -   Prettier <3.4 used to add extra spaces after the list item prefix.
    -   Prettier 3.4 will now remove the extra spaces.

!!! note "Example of an admonition"

    Text content inside an admonition is indented by one tab width.

!!! example "Example of an admonition containing a code block"

    ```sh
    echo \
      "When code blocks are inside un-nested admonitions," \
      "Prettier seems to be behaving as expected."
    ```

-   Admonitions can be nested within lists.
-   Like this:

    !!! example "Example of a nested admonition containing a code block"

        ```sh
        echo \
          "When code blocks are inside nested admonitions," \
          "Prettier seems to be re-indenting code blocks every time it runs."
        ```

````

Renders like this using Material for MkDocs:

<img width="782" alt="prettier-prettier-16929-issue-material-for-mkdocs-example-formatted-by-prettier-3-3" src="https://github.com/user-attachments/assets/dd743258-3b3b-4098-bd2f-74755fa25994" />

Running Prettier over the file like this:

```sh
~/path/to/example
❯ prettier --version
3.4.2

~/path/to/example
❯ prettier example.md --log-level debug --write
[debug] normalized argv: {"":["example.md"],"cache":false,"color":true,"editorconfig":true,"write":true,"logLevel":"debug","configPrecedence":"cli-override","debugRepeat":0,"ignorePath":[".gitignore",".prettierignore"],"plugins":[],"_":["example.md"],"__raw":{"_":["example.md"],"cache":false,"color":true,"editorconfig":true,"write":true,"log-level":"debug","config-precedence":"cli-override","debug-repeat":0,"ignore-path":[".gitignore",".prettierignore"],"plugin":[]}}
[debug] resolve config from '/path/to/example/example.md'
[debug] loaded options `{"tabWidth":4}`
[debug] applied config-precedence (cli-override): {"tabWidth":4}
example.md 21ms
```

Results in Markdown content like this:

````
# Example

- This is an unordered list.
    - Prettier <3.4 used to add extra spaces after the list item prefix.
    - Prettier 3.4 will now remove the extra spaces.

!!! note "Example of an admonition"

    Text content inside an admonition is indented by one tab width.

!!! example "Example of an admonition containing a code block"

    ```sh
    echo \
      "When code blocks are inside un-nested admonitions," \
      "Prettier seems to be behaving as expected."
    ```

- Admonitions can be nested within lists.
- Like this:

    !!! example "Example of a nested admonition containing a code block"

        ```sh
        echo \
          "When code blocks are inside nested admonitions," \
          "Prettier seems to be re-indenting code blocks every time it runs."
        ```

````

The plain unordered lists format correctly. Prettier 3.3 added three spaces before the `-`, but Prettier 3.4 will reduce that to a single space.

But `prettier --check` says there are still formatting issues:

```sh
~/path/to/example
❯ prettier example.md --log-level debug --check
[debug] normalized argv: {"":["example.md"],"cache":false,"check":true,"color":true,"editorconfig":true,"logLevel":"debug","configPrecedence":"cli-override","debugRepeat":0,"ignorePath":[".gitignore",".prettierignore"],"plugins":[],"_":["example.md"],"__raw":{"_":["example.md"],"cache":false,"check":true,"color":true,"editorconfig":true,"log-level":"debug","config-precedence":"cli-override","debug-repeat":0,"ignore-path":[".gitignore",".prettierignore"],"plugin":[]}}
Checking formatting...
[debug] resolve config from '/path/to/example/example.md'
[debug] loaded options `{"tabWidth":4}`
[debug] applied config-precedence (cli-override): {"tabWidth":4}
[warn] example.md
[warn] Code style issues found in the above file. Run Prettier with --write to fix.
```

Running `prettier --write` _a second time_ after the list formatting has been corrected results in Markdown content like this:

````
# Example

- This is an unordered list.
    - Prettier <3.4 used to add extra spaces after the list item prefix.
    - Prettier 3.4 will now remove the extra spaces.

!!! note "Example of an admonition"

    Text content inside an admonition is indented by one tab width.

!!! example "Example of an admonition containing a code block"

    ```sh
    echo \
      "When code blocks are inside un-nested admonitions," \
      "Prettier seems to be behaving as expected."
    ```

- Admonitions can be nested within lists.
- Like this:

    !!! example "Example of a nested admonition containing a code block"

          ```sh
          echo \
            "When code blocks are inside nested admonitions," \
            "Prettier seems to be re-indenting code blocks every time it runs."
          ```

````

Prettier adds two spaces to the indentation of the code block (which is not even the `tabWidth` in this case), _every time it runs_. This also means `prettier --check` will fail every time it runs.

Prettier ignore comments (`<!-- prettier-ignore -->`) do not prevent this behavior.

**Comment by ferrarimarco** [comment]
Hi.

It seems that #15526 introduced a bug according to the CommonMark spec, and also produces documents that don't honor the `tabwidth` setting.

The CommonMark spec says this about [list items](https://spec.commonmark.org/0.31.2/#list-items):

> [...] If a sequence of lines Ls constitute a sequence of blocks Bs starting with a character other than a space or tab, and M is a list marker of width W followed by 1 ≤ N ≤ 4 spaces of indentation, then the result of prepending M and the following spaces to the first line of Ls, and indenting subsequent lines of Ls by W + N spaces, is a list item with Bs as its contents. [...]

and

> [...] The most important thing to notice is that the position of the text after the list marker determines how much indentation is needed in subsequent blocks in the list item. If the list marker takes up two spaces of indentation, and there are three spaces between the list marker and the next character other than a space or tab, then blocks must be indented five spaces in order to fall under the list item. [...]

And it says this about [ordered list markers](https://spec.commonmark.org/0.31.2/#ordered-list-marker):

> An ordered list marker is a sequence of 1–9 arabic digits (0-9), followed by either a . character or a ) character. (The reason for the length limit is that with 10 digits we start seeing integer overflows in some browsers.)

So, we have:

```
W('*') = 1
```

So, the allowed indentation (I) according to the CommonMark spec (this is the number of spaces between the list marker of a list item and the content of that list item):

```
I('*') = W('*') + N = 1 + N
```

When Prettier <3.4.0 formats unordered lists, it can automatically infer the value of N using the `tabWidth` setting (and thus the value of I):

- `tabwidth: 2`-> Prettier sets N=1 -> I('*') = 2
- `tabwidth: 4` -> Prettier sets N=3 -> I('*') = 4

```markdown
<!-- tabWidth = 2 --!>
- Long item test testtesttesttesttesttesttesttest
  test
- Item 2

<!-- tabWidth = 4 --!>
-   Long item test testtesttesttesttesttesttesttest
    test
-   Item 2
```

When Prettier >=3.4.0 formats unordered lists, it produces a document that's not in line with the CommonMark spec, and also not honoring `tabwidth`:

- `tabwidth: 2`-> Prettier sets N=1 -> I('*') = 2
- `tabwidth: 4` -> Prettier sets N=1 -> I('*') = 2

```markdown
<!-- tabWidth = 2 --!>
- Long item test testtesttesttesttesttesttesttest
  test
- Item 2

<!-- tabWidth = 4 --!>
- Long item test testtesttesttesttesttesttesttest
  test <!-- violates tabwidth=4 --!>
- Item 2
```

**Comment by 8Crafter** [comment]
I have been using Prettier v3.3.3 for so long now to avoid this issue, but the outdated parsers/printers for other languages that come as a consequence of using that older version have been really annoying, so I made a plugin that reverts this change, so I can use the latest Prettier versions while still having the old list formatting behavior (the plugin just replaces the Markdown printer with the one from Prettier v3.3.3): https://www.npmjs.com/package/prettier-plugin-markdown-list-tabwidth

For anyone else using Prettier v3.3.3 to avoid this, I recommend using this plugin until Prettier adds a built-in way to revert this.

**Comment by enduity** [comment]
I think code blocks even more obviously illustrate the problem that has been created. I used `tabWidth: 4` here.

Input:
````markdown
- Fusce dictum metus eget posuere efficitur. <br/>
  Morbi vitae mauris ipsum.
  ```shell
  git add .
  ```
````

Output as of Prettier 3.8.1 (width `tabWidth` of 4):
````markdown
- Fusce dictum metus eget posuere efficitur. <br/>
  Morbi vitae mauris ipsum.
    ```shell
    git add .
    ```
````

The alignment is inconsistent, making the resulting output hard to read.

<details>
<summary>More complex example</summary>

Input (how I would personally format it):
````markdown
1.  Lorem ipsum dolor sit amet, consectetur adipiscing elit. <br/>
    Fusce interdum egestas diam, quis accumsan nulla mollis nec.
    -   Fusce dictum metus eget posuere efficitur.
        Morbi vitae mauris ipsum.

        Aliquam erat volutpat.

        ```shell
        git add .
        ```

        -   Donec auctor.
        -   Sed tellus nibh, aliquam at sagittis id.

    -   Fusce nisl elit, scelerisque vitae volutpat porta.
````

Output as of Prettier 3.8.1 (width `tabWidth` of 4):
````markdown
1.  Lorem ipsum dolor sit amet, consectetur adipiscing elit. <br/>
    Fusce interdum egestas diam, quis accumsan nulla mollis nec.
    - Fusce dictum metus eget posuere efficitur.
      Morbi vitae mauris ipsum.

        Aliquam erat volutpat.

        ```shell
        git add .
        ```

        - Donec auctor.
        - Sed tellus nibh, aliquam at sagittis id.

    - Fusce nisl elit, scelerisque vitae volutpat porta.
````

In the render, "Aliquam erat volutpat." is aligned with "Fusce dictum metus eget posuere efficitur.", as it was intended. There are three different levels of text indentation.

However, in the Prettier-formatted markdown, there are now four different levels of text indentation. This is hard to follow with your eyes.
</details>

**Comment by kareldonk** [comment]
Is this fucking retarded change going to be fixed/rolled back or what?
